File: cpp/347_top_k_frequent_elements.cpp
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  void countFrequency(unordered_map<int, int> &num_freq,
                      const vector<int> &nums) {
    for (const auto &num : nums) {
      num_freq[num]++;
    }
  }

  void fillBuckets(vector<vector<int>> &buckets,
                   const unordered_map<int, int> &num_freq) {
    for (const auto &[key, value] : num_freq) {
      buckets[value - 1].push_back(key);
    }
  }

  vector<int> topKFrequent(vector<int> &nums, int k) {
    unordered_map<int, int> num_freq;
    countFrequency(num_freq, nums);

    vector<vector<int>> buckets(nums.size());
    fillBuckets(buckets, num_freq);

    vector<int> result;
    for (int i = buckets.size() - 1; i > -1; --i) {
      for (const auto &num : buckets[i]) {
        result.push_back(num);
        if (result.size() == k) {
          return result;
        }
      }
    }
    return result;
  }
};

File: rust/347_top_k_frequent_elements.rs
impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        use std::collections::HashMap;

        let mut num_freq = HashMap::new();
        let mut max_freq: usize = 0;
        nums.into_iter().for_each(|num| {
            let entry = num_freq.entry(num).or_default();
            *entry += 1;
            max_freq = std::cmp::max(max_freq, *entry);
        });

        let mut buckets: Vec<Vec<i32>> = vec![vec![]; max_freq + 1];
        for (num, freq) in num_freq {
            buckets[freq].push(num);
        }

        let mut result = vec![];
        for i in (0..max_freq + 1).rev() {
            for num in &buckets[i] {
                result.push(*num);
                if result.len() == k as usize {
                    return result;
                }
            }
        }

        result
    }
}
